fix #14: make `resolve` synchronous

giltayar opened the pull request:
As described in issue #14, `import.meta.resolve` is becoming synchronous both in the browser and in Node.js:

You can see it here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import.meta/resolve

And also in Node.js, where a PR just changed it to be synchronous: nodejs/node#44710

This PR makes `resolve` synchronous to fit the above. 

Note that the baseline test calling `import.meta.resolve` was _not_ changed to be synchronous,
because that change will probably be  merged into Node v19, 
and so for 18 `import.meta.resolve` is still synchronous.

wooorm commented on index.js:
```diff
@@ -15,11 +15,11 @@ import {defaultResolve} from './lib/resolve.js'
- * @returns {Promise<string>}
- *   Returns a promise that resolves to a full `file:`, `data:`, or `node:` URL
+ * @returns {string}
+ *   Returns a string that resolves to a full `file:`, `data:`, or `node:` URL
```
```suggestion
 *   Returns a string to a full `file:`, `data:`, or `node:` URL
```

wooorm commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "import-meta-resolve",
-  "version": "2.2.2",
+  "version": "3.0.0",
```
```suggestion
  "version": "2.2.2",
```

I’ll do this in a separate commit!

wooorm commented on test/core.js:
```diff
@@ -14,6 +14,8 @@ const windows = process.platform === 'win32'
 const veryOldNode = semver.lt(process.versions.node, '16.0.0')
 const oldNode = semver.lt(process.versions.node, '18.0.0')
 
+const execute = (/** @type {() => void} */ f) => f()
```
Maybe a comment above this to say that it will be changed by the build script to an async version?

wooorm replied:
Also: I somewhat prefer the term `run`: a) it’s shorter, b) it’s not a synonym for killing things. What do you think?

giltayar replied:
I'm less inclined to leave a comment about the function being changed to async because there's a lot of sync/async transformations going on in this file. I would prefer adding a comment in the build script of why we're doing this.

giltayar replied:
Added a comment in `script.js`

wooorm commented on script.js:
```diff
@@ -4,7 +4,13 @@ import fs from 'node:fs/promises'
+    'const execute = async (/** @type {() => Promise<void>} */ f) => f()'
+  )
+  .replace(/execute\(/g, 'await execute(async ')
   .split('\n')
```
Should we perhaps generate two baseline test files, one async one sync, and a runtime file that checks the current Node version, and imports one or the other?

giltayar replied:
I actually thought of doing this, but there is currently _no_ version of Node.js that supports the sync version! I will try and remember, once one comes out, of doing this. And also cleaning up some out of date code that seems to deal with _really_ old versions of Node.js (I _think_ I saw something there).

wooorm reviewed (commented):
Thanks Gil for working on this! Much appreciated!

Some notes!

wooorm commented:
Thinking more about https://github.com/wooorm/import-meta-resolve/pull/15/files#r1165850902, how strongly do you feel about releasing this now?

Given that: a) this isn’t supported anywhere yet, b) this is a major, c) Node 16 is dropped at the end of the month, we could also wait until April 30, and drop Node 16 too?

giltayar commented:
I think waiting until end of April and dropping support of 16 is a good idea. And maybe release this only when the change really _is_ in Node and we can check the baseline against it. BTW, my guess is that this change to `import.meta.resolve` won't be backported to Node 18 anyway, so we'll have to live with dual sync and async `import.meta.resolve` in Node.js anyway.

Note that while this _is_ a breaking change, it won't _really_ break any code, given that `await import.meta.resolve(...)` works even when the function doesn't return a promise. It _might_ break some linter/TS rules, but that's about it.

TL;DR. I suggest we wait for two things:

1. Drop of Node 16 at end of April
2. Checking this against the version of Node that implemented this change, so we can implement two baselines test: one for version 18 (async) and one for version 20 (sync). I'm closely following this, so I can both know when it happens and implement the change.

giltayar commented:
Node 20 is out and `import.meta.resolve` is now synchronous on it, and can be checked! Since this PR is still open, I'll continue it and create two baselines: an async one for "Node <20' and a sync one for 'Node>20` as we discussed above.

giltayar commented:
I implemented (and manually tested on 16, 18, 20) the synchronous baseline test, and added an async one that runs for versions < 20.

I changed the action `main.yml` to test on `18` and `latest`, but I wasn't sure about that, so it's a separate commit.

wooorm commented on .github/workflows/main.yml:
```diff
@@ -20,5 +20,5 @@ jobs:
-          - lts/hydrogen
-          - node
+          - 18
+          - latest
```
```suggestion
          - node
```

Not sure about the differences between the two, but `node` seems a bit more stable as a label than `latest`.
Anyway, that’s what it used to be, we could change it but that’s probably a different discussion about trade offs!

giltayar replied:
Yeah, the documentation says that `latest` and `node` are the same, and `latest` sounded clearer to me. But I'm fine with keeping it as it is.

wooorm commented on test/core.js:
```diff
@@ -3,625 +3,601 @@
 process.on('exit', async () => {
   // Has to be sync.
-  renameSync('package.json.bak', 'package.json')
+  if (existsSync('package.json.bak')) {
```
I’d prefer not to check this, but instead just try/catch it. AFAIK the docs still say that `exists*` and then doing something is an anti-pattern?

giltayar replied:
Yeah, I know. It's in the docs, and also the reason why there's no `existsAsync`. But I really don't get the argument, and in the end it's irrelevant in this case because there are no multiple processes dealing with this file.

I can do it using a `try/catch`, but I think this is clearer. Your call. I prefer the existing code, but just say the word, and it changes to `try/catch`! 😊

wooorm replied:
Yes, I would prefer just reading.
Either a simple one:

```
try {
  renameSync('package.json.bak', 'package.json')
} catch {}
```

Or a more involved:

```
try {
  renameSync("package.json.bak", "package.json");
} catch (error) {
  const exception = /** @type {ErrnoException} */ (error)
  // this next line is pseudo code: not sure what the exact code is off the top of my head
  if (exception.code === 'ERR_NOT_FOUND') return
  throw exception
}
```

giltayar replied:
Used the more involved one.😊

wooorm commented on test/core.js:
```diff
@@ -3,625 +3,601 @@
-  assert(resolve, 'expected `resolve` to exist (needed for TS in baseline)')
+test(
+  'resolve(specifier, base?, conditions?)',
+  {skip: false},
```
Is this line a left over from some local testing?

giltayar replied:
Ah, no. Maybe I should put in a comment about this. 

Look at `script.js`, which modifies `baseline.js` to change `fskip: alse` to "skip if version < 20", i.e. skip checking the sync version of `import.meta.resolve` if v<20.

I tried just skipping the whole test (using an `if (...) process.exit(0)` at the beginning of the file), but the coverage raised an alarm about not enough coverage. Only skipping the test worked. Not sure why...

The alternative for replacing the `skip` would be for `script.js` to generate the sync baseline test only if v > 20, but I didn't like that for some reason. Can't exactly point the finger to why.

wooorm replied:
Oh smart Hmm, sad about the coverage lines.

A comment is a good idea! Something like this?
```
// Note: this is toggled by the reference tests.
```

giltayar replied:
Added comment!

wooorm reviewed (commented):
Really appreciate your continued work Gil! Some tiny nits and then it’s ready I think!

wooorm commented:
Yes!
I don’t think I need to wait for April 30, for Node 16 EOL, I can also just do that now.

Anything else you want to get in the release?

wooorm commented:
Released in https://github.com/wooorm/import-meta-resolve/releases/tag/3.0.0!

nicolo-ribaudo commented:
Thank you!

wooorm commented:
I made a mistake with 16 being EOL, that is supposed to be in a year, but is instead in 6 months.

However, Node 14 is EOL this week. That is intentional

wooorm commented:
@jakebailey Given that 14 is EOL in 5 days, why do you think this project should support that for the foreseeable future?

(reference: https://github.com/nodejs/Release#release-schedule)

wooorm commented:
The problem for me is that there aren’t really good reasons to switch the code style to depend on a new Node version, for example, `Object.hasOwn` could be a reason, but well that isn’t *super* useful. That’s not worth a major.

But there are good reasons to switch the infra to depend on a new Node version. E.g., `node:test` or so.
So it would be nice for the tooling in this project to work in maintained versions, but not be limited to unmaintained versions.

Hence why I am *currently* thinking that the best way to go about it is, if a major is going to happen anyway, to drop support for unmaintained Node versions. And describe that as such. Even if the code might *currently* work, it allows for future patch/minor releases with new/changed dependencies or more modern syntax that only work in those versions too.

Open to ideas!

I’ve updated the readme/actions to check Node 16 again, and amended the release notes to specify Node 16 too.

nicolo-ribaudo commented:
Babel already has a build step around this library, so the choice here doesn't really affect our choice.

It's possible that we'll keep Node.js 14 compatibility, in which case we'll just transpile usages of JS functions introduced in Node.js 16 to something that works in Node.js 14.
